File: ArrayAndHashing/ContainerWithMostWater/cwmw.go
package main

import "fmt"

func maxArea(height []int) int {
	if len(height) <= 0 {
		return 0
	}
	m := 0
	l := 0
	r := len(height) - 1
	for l < r {
		w := r - l
		lh := height[l]
		rh := height[r]
		h := min(lh, rh)
		if m < w*h {
			m = w * h
		}
		if lh < rh {
			l++
		} else {
			r--
		}
	}
	return m
}

func main() {
	s1 := []int{1, 8, 6, 2, 5, 4, 8, 3, 7}
	fmt.Printf("T1: %d\n", maxArea(s1))

	s2 := []int{1, 1}
	fmt.Printf("T1: %d\n", maxArea(s2))
}

File: TwoPointers/ContainerWithMostWater/cwmw.py
from typing import List


class Solution:
    def maxArea(self, heights: List[int]) -> int:
        res, l, r = 0, 0, len(heights) - 1

        while l < r:
            width = r - l
            res = max(res, width * min(heights[l], heights[r]))
            if heights[l] <= heights[r]:
                l += 1
            else:
                r -= 1

        return res


if __name__ == "__main__":
    print("MAIN:")
    sol = Solution()

    h = [1, 7, 2, 5, 4, 7, 3, 6]
    r = sol.maxArea(h)
    print(f"T1: {r}")

    h = [2, 2, 2]
    r = sol.maxArea(h)
    print(f"T2: {r}")
